notebooks/network-analysis.ipynb

```diff
@@ -0,0 +1,434 @@
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Identify reaches\n",
+    "- Identifying reaches in each subnetwork is a two step process\n",
+    "    1. Use the `reachable_network` function to identify the upstream connectivity structure of each subnetwork. The function accepts the reverse network dictionary, `rconn`, and returns a dictionary, `subnets_`, which lists upstream connections for each segment in each subnetwork. \n",
+    "    2. Conduct a [depth-first-search](https://en.wikipedia.org/wiki/Depth-first_search) of each subnetwork. This is done with the `dfs_decomposition` function, which accepts the forward (downstream oriented) network dictionary, `nhd_network.reverse_network(net)`, and returns a dictionary of segment lists defining specific reaches in the subnetwork. Results from each subnetwork are concatenated into a larger dictionary, `subreaches`, which contains segment lists for all reaches in the supernetwork.    \n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create reverse connection dictionary for each subnetwork\n",
+    "# tailwater-segment-id: {segment-id: [upstream-connected-segment-id]}\n",
+    "subnets_ = nhd_network.reachable_network(rconn)\n",
+    "\n",
+    "# create a dictionary, `subreaches` containing lists of all linear connected segments between junctions (reaches) in the supernetwork. \n",
+    "# Recall, a reach is composed of several interconnected segments between junctions.  \n",
+    "# tailwater-segment-id: [[rch1-segment-id-1, rch1-segment-id-2, ..., rch1-segment-id-n],[rch2-segment-id1, rch2-segment-id2, ..., rch2-segment-id-n]]\n",
+    "subreaches = {}\n",
+    "for tw, net in subnets_.items(): # for the number of tailwaters in the supernetwork\n",
+    "    \n",
+    "    # I dont know what this does... something with identifying junctions\n",
+    "    path_func = partial(nhd_network.split_at_junction, net)\n",
+    "    \n",
+    "    # conduct a depth-first search to decompose the network into list of reaches\n",
+    "    reach = nhd_network.dfs_decomposition(\n",
+    "        nhd_network.reverse_network(net), path_func\n",
+    "    )\n",
+    "    \n",
+    "    # add reach list for subnetwork tw to the subreaches library\n",
+    "    subreaches[tw] = reach"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Visualize reaches in a single subnetwork\n",
+    "- Make a map showing each of the reaches in the Colorado River Basin subnetwork"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# extract a single large subnetwork from the supernetwork GeoDataFrame.\n",
+    "# Turns out that subnetwork code 210 is the Colorado River Basin\n",
+    "C = gdf[\"subnetwork\"] == 210\n",
```

Comment:
There should be a more robust way to do this. It's never guaranteed to always be code 210.